pkg/cached/cached.go

```diff
@@ -2,368 +2,598 @@ package cached
+		return fmt.Errorf("failed to set permission on base lv %s: %w", lv, err)
 	}
 
-	logger.Info(ctx, "volume unpublished and data removed", key.TargetPath.Field(targetPath))
-	return &csi.NodeUnpublishVolumeResponse{}, nil
+	logger.Info(ctx, "deactivating base lv", key.LV.Field(lv))
+	if err := exec.Run(ctx, "lvchange", "--activate", "n", lv); err != nil {
+		return fmt.Errorf("failed to deactivate base lv %s: %w", lv, err)
+	}
+
+	logger.Info(ctx, "prepared base pv",
+		key.PV.Field(c.BasePV),
+		key.VG.Field(vg),
+		key.LV.Field(lv),
+		key.Device.Field(lvDevice),
+	)
+
+	return nil
 }
 
-// NodeGetVolumeStats returns the volume capacity statistics available for the given volume.
-func (c *Cached) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	if req.VolumeId == "" {
-		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume ID must be provided")
+func (c *Cached) importBasePV(ctx context.Context) error {
+	ctx = logger.With(ctx, key.PV.Field(c.BasePV), key.VG.Field(c.VG))
+	ctx, span := telemetry.Start(ctx, "cached.import-base-pv")
+	defer span.End()
+
+	err := exec.Run(ctx, "vgdisplay", c.VG)
+	if err == nil {
+		logger.Info(ctx, "vg already exists")
+		return nil
 	}
 
-	volumePath := req.VolumePath
-	if volumePath == "" {
-		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume Path must be provided")
+	if !strings.Contains(err.Error(), "not found") {
+		return fmt.Errorf("failed to check vg %s: %w", c.VG, err)
 	}
 
-	usedBytes, err := getFolderSize(volumePath)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to retrieve used size statistics for volume path %s: %v", volumePath, err)
+	logger.Info(ctx, "importing base pv")
+	if err := exec.Run(ctx, "vgimportclone", "-n", c.VG, c.BasePV); err != nil {
+		return fmt.Errorf("failed to import base pv %s: %w", c.BasePV, err)
 	}
 
-	var stat syscall.Statfs_t
-	err = syscall.Statfs(volumePath, &stat)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to retrieve total size statistics for volume path %s: %v", volumePath, err)
-	}
-
-	// Calculate free space in bytes
-	freeBytes := stat.Bavail * uint64(stat.Bsize)
-	if freeBytes > math.MaxInt64 {
-		return nil, status.Errorf(codes.Internal, "total size statistics for volume path too big for int64: %d", freeBytes)
-	}
-	signedFreeBytes := int64(freeBytes)
-
-	return &csi.NodeGetVolumeStatsResponse{
-		Usage: []*csi.VolumeUsage{
-			{
-				Available: signedFreeBytes,
-				Total:     signedFreeBytes + usedBytes,
-				Used:      usedBytes,
-				Unit:      csi.VolumeUsage_BYTES,
-			},
-		},
-	}, nil
-}
+	logger.Info(ctx, "extending vg with thinpool pvs", key.DeviceGlobs.Field(c.ThinpoolPVGlobs), key.PVs.Field(c.ThinpoolPVs))
+	if err := exec.Run(ctx, "vgextend", append([]string{"--config=devices/allow_mixed_block_sizes=1", c.VG}, c.ThinpoolPVs...)...); err != nil {
```

Comment:
Do we need the mixed block sizes? I think I saw something about it affecting performance negatively.